### DISCUSSION TITLE
Sixel image quality

### DISCUSSION THREAD

**Discussion by ghost** [hpjansson/chafa#173] [Q&A] [answered]
I intend to use chafa as a previewer for my photos, but have sometimes problems with the rendering quality of slow color gradients (e.g. cloudy skies), that come out very quantized with distinct color bands. Crisp photos render however perfectly as sixels. I have tried chafa, notcurses, img2sixel and imagemagick and in general I can get the best results with the -I option of img2sixel, followed by magick, notcurses and chafa (my terminal is foot version: 1.14.0 +pgo +ime +graphemes -assertions from nixos 23.05). Here are the resulting sixel files, png images have been generated with sixel2img:

Original (not very beautiful, but shows the effect very clearly):

![](https://files.pdp8.info/source.jpeg)

`chafa source.jpeg > chafa.six` (Chafa version 1.12.4)

![](https://files.pdp8.info/chafa.png)
<https://files.pdp8.info/chafa.six>

`ncplayer -t0 -q -k -s scalehi source.jpeg > notcurses.six` (ncplayer version 3.0.9)

![](https://files.pdp8.info/notcurses.png)
<https://files.pdp8.info/notcurses.six>

`magick source.jpeg -resize 1914x1188>  sixel:- > magick.six` (Version: ImageMagick 7.1.1-18 Q16-HDRI x86_64 8c6cdf58e:20230923 https://imagemagick.org)

![](https://files.pdp8.info/magick.png)
<https://files.pdp8.info/magick.six>

`img2sixel -I -w 1914 source.jpeg > img2sixel.six` (img2sixel 1.10.3)

Conversion to png spoils it completely, but img2sixel gives by far the best approximation of the original with the `-I` option.
The high color option may cause however artefacts at the bottom of some images (maybe related to <https://github.com/libsixel/libsixel/issues/16>).

![](https://files.pdp8.info/img2sixel.png)
<https://files.pdp8.info/img2sixel.six>

Are there any options in chafa to improve sixel rendering of photos? Dithering can improve the appearance of this file, but I loose details in sharper shots. I would prefer to use chafa over the other options because of speed and active maintainance.

**Comment by hpjansson** [comment]
The `-w` option is supposed to govern the speed/quality tradeoff, but unfortunately it doesn't have any effect on sixel quality yet. There's a lot that could be done to improve it, though -- for instance, we're probably allocating a disproportionate amount of palette entries to the most common color ranges (cf. the dark areas), and too few to rarer colors. Thanks for the nice comparison pics -- I'll create an issue to work on it.

Edit: https://github.com/hpjansson/chafa/issues/174

**Comment by ghost** [comment]
Thanks a lot for looking into it! Please let me know, if you need some testing.

**Comment by hpjansson** [comment, marked as answer]
I'm a bit sad that the original reporter disappeared, because this is now [thoroughly fixed](https://github.com/hpjansson/chafa/issues/174#issuecomment-2489241676)!
